Reuse already-installed images when doing an upgrade

saiarcot895 opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/sonic-net/SONiC/blob/gh-pages/CONTRIBUTING.md

Please provide following information to help code review process a bit easier:
-->
### Description of PR
<!--
- Please include a summary of the change and which issue is fixed.
- Please also include relevant motivation and context. Where should reviewer start? background context?
- List any dependencies that are required for this change.
-->

Summary:
Fixes # (issue)

### Type of change

<!--
- Fill x for your type of change.
- e.g.
- [x] Bug fix
-->

- [ ] Bug fix
- [x] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)


### Back port request
- [ ] 201911
- [ ] 202012
- [ ] 202205
- [ ] 202305
- [ ] 202311

### Approach
#### What is the motivation for this PR?

For `test_upgrade_path.py`, if the from image or the to image is already installed, then instead of removing it and reinstalling it, reuse the existing image by restoring it to the pre-first-boot state (so that when it's booted into, it'll be like it's the first time). This saves some time in downloading and installing the image file to this device.

For this to work, the image version needs to be passed in as a parameter; otherwise, the version is not known, and the existing behavior of removing and reinstalling the image will apply.

#### How did you do it?

#### How did you verify/test it?

#### Any platform specific information?

#### Supported testbed topology if it's a new test case?

### Documentation
<!--
(If it's a new feature, new test case)
Did you update documentation/Wiki relevant to your implementation?
Link to the wiki page?
-->

saiarcot895 committed:
Reuse already-installed images when doing an upgrade

For `test_upgrade_path.py`, if the from image or the to image is already
installed, then instead of removing it and reinstalling it, reuse the
existing image by restoring it to the pre-first-boot state (so that when
it's booted into, it'll be like it's the first time). This saves some
time in downloading and installing the image file to this device.

For this to work, the image version needs to be passed in as a
parameter; otherwise, the version is not known, and the existing
behavior of removing and reinstalling the image will apply.

Signed-off-by: Saikrishna Arcot <sarcot@microsoft.com>

mssonicbld commented:
The [pre-commit](http://pre-commit.com/) check detected issues in the files touched by this pull request.
The pre-commit check is a mandatory check, please fix detected issues.

Detailed pre-commit check results:
<samp>trim trailing whitespace.................................................Passed<br/>fix end of files.........................................................Passed<br/>check yaml...........................................(no files to check)Skipped<br/>check for added large files..............................................Passed<br/>check python ast.........................................................Passed<br/>flake8...................................................................Failed<br/>- hook id: flake8<br/>- exit code: 1<br/><br/>tests/upgrade_path/conftest.py:5:18: F821 undefined name 'request'<br/>tests/upgrade_path/conftest.py:6:16: F821 undefined name 'request'<br/>tests/upgrade_path/test_upgrade_path.py:60:121: E501 line too long (127 > 120 characters)<br/>tests/upgrade_path/upgrade_helpers.py:77:17: F821 undefined name 'os'<br/>tests/upgrade_path/upgrade_helpers.py:78:36: F821 undefined name 'os'<br/>tests/upgrade_path/upgrade_helpers.py:80:121: E501 line too long (122 > 120 characters)<br/><br/>flake8...............................................(no files to check)Skipped<br/>check conditional mark sort..........................(no files to check)Skipped<br/></samp>

To run the pre-commit checks locally, you can follow below steps:
1. Ensure that default python is python3. In sonic-mgmt docker container, default python is python2. You can run
   the check by activating the python3 virtual environment in sonic-mgmt docker container or outside of sonic-mgmt
   docker container.
2. Ensure that the `pre-commit` package is installed:
```
sudo pip install pre-commit
```
3. Go to repository root folder
4. Install the pre-commit hooks:
```
pre-commit install
```
5. Use pre-commit to check staged file:
```
pre-commit
```
6. Alternatively, you can check committed files using:
```
pre-commit run --from-ref <commit_id> --to-ref <commit_id>
```

saiarcot895 committed:
Fix pre-commit errors

Signed-off-by: Saikrishna Arcot <sarcot@microsoft.com>

saiarcot895 committed:
Convert back --base_image/--target_image to the list versions

Other test cases appear to be using --base_image_list and/or
--target_image_list. Because of that, change back to using those
arguments, and convert the image version argument to list versions.

Right now, the upgrade path test case doesn't actually accept lists, but
support for this will be added in a future commit.

Signed-off-by: Saikrishna Arcot <sarcot@microsoft.com>

saiarcot895 committed:
Add support for multiple images specified in base and/or target image

This will run the specified task individually for each combination of
base and target image. Specifically, it will do a cold reboot to the
base image, and then cold/warm/fast reboot to the target image.

Signed-off-by: Saikrishna Arcot <sarcot@microsoft.com>

prgeor commented on tests/upgrade_path/restoreToFirstBoot.sh:
```diff
@@ -0,0 +1,20 @@
+docker system prune -f
+
+rm -rf /host/${IMAGE_DIR_NAME}/rw
+touch /host/${IMAGE_DIR_NAME}/platform/firsttime
```
@saiarcot895 what are the things today taken care of ONLY during firstime?

saiarcot895 replied:
Installation of the platform-specific support package, and copying configs from `/host/old_config` to `/etc/sonic`. I'm not sure if there's others.

prgeor commented on tests/metadata-scripts/getImageHash.py:
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env python
```
@saiarcot895 where is this file used?

saiarcot895 replied:
Not used here, removed the file.

prgeor commented on tests/upgrade_path/upgrade_helpers.py:
```diff
@@ -59,6 +60,32 @@ def check_sonic_version(duthost, target_version):
         "Upgrade sonic failed: target={} current={}".format(target_version, current_version)
 
 
+def restore_image_to_first_boot(duthost, image_version):
```
@saiarcot895 define the function helper what it does and what it return value mean?

saiarcot895 replied:
Added description.

saiarcot895 committed:
Remove unused file

Signed-off-by: Saikrishna Arcot <sarcot@microsoft.com>

saiarcot895 committed:
Add description for restore_to_first_boot function

Signed-off-by: Saikrishna Arcot <sarcot@microsoft.com>

vaibhavhd commented on tests/upgrade_path/upgrade_helpers.py:
```diff
@@ -59,6 +60,40 @@ def check_sonic_version(duthost, target_version):
         "Upgrade sonic failed: target={} current={}".format(target_version, current_version)
 
 
+def restore_image_to_first_boot(duthost, image_version):
```
This new mechanism should be optionally enabled. For folks in community who still want to use the old way (without the hack here).

saiarcot895 replied:
This mechanism will only work if `--base_image_version_list` or `--target_image_version_list` are passed in. Otherwise, current behavior is used.

vaibhavhd commented on tests/upgrade_path/upgrade_helpers.py:
```diff
@@ -59,6 +60,40 @@ def check_sonic_version(duthost, target_version):
+    base_path = os.path.dirname(__file__)
+    restore_to_first_boot_script = os.path.join(base_path, "restoreToFirstBoot.sh")
+    duthost.copy(src=restore_to_first_boot_script, dest="/tmp/restoreToFirstBoot.sh", mode="preserve")
+    duthost.command("sudo unshare -mnpf /tmp/restoreToFirstBoot.sh {}"
```
Double checking - will the logic `image_version.replace("SONiC-OS", "image")` work for all versions (public, internal, tagged?)

saiarcot895 replied:
Yes. For example, public is something like `SONiC-OS-master.602773-0324d9acd`.

vaibhavhd commented on tests/upgrade_path/upgrade_helpers.py:
```diff
@@ -59,6 +60,40 @@ def check_sonic_version(duthost, target_version):
+                    .format(image_version.replace("SONiC-OS", "image")))
+    duthost.command("sudo sonic-installer set-next-boot {}".format(image_version))
+    duthost.file(path="/host/old_config", state="absent")
+    duthost.copy(src="/etc/sonic/", dest="/host/old_config/", remote_src=True)
```
I think `rm "/host/old_config/config_db.json"` is also needed. We don't want any kind of reboot (cold, warm, fast) to use config from old_config, rather use minigraph as source of truth.

vaibhavhd commented on tests/upgrade_path/restoreToFirstBoot.sh:
```diff
@@ -0,0 +1,20 @@
+
+dockerd -l warn &
+
+sleep 3
```
Why sleep?

vaibhavhd commented on tests/upgrade_path/test_upgrade_path.py:
```diff
@@ -53,24 +81,30 @@ def cleanup_prev_images(duthost):
+    logger.info("Test upgrade path from {} ({}) to {} ({})".format(from_image, from_image_version,
+                                                                   to_image, to_image_version))
+    target_version = None
+    if from_image_version and restore_image_to_first_boot(duthost, from_image_version):
```
This change will help in not installing the base OS which may already be installed.

Is it intentional to not make the change to install the target OS on DUT if it is already installed?

vaibhavhd commented on tests/platform_tests/args/advanced_reboot_args.py:
```diff
@@ -126,13 +126,25 @@ def add_advanced_reboot_args(parser):
+        )
+
+    parser.addoption(
+        "--target_image_version_list",
```
what is the diff now between target_image_list and target_image_version_list? Seems redundant.

saiarcot895 replied:
`--target_image_list` is a comma-separated list of URLs pointing to SONiC images. `--target_image_version_list` is a comma-separated list of SONiC versions (mapped to the URLs that are passed in).

saiarcot895 committed:
Merge remote-tracking branch 'origin/master' into reuse-existing-image-in-upgrade

Signed-off-by: Saikrishna Arcot <sarcot@microsoft.com>

saiarcot895 committed:
Don't get nbrhosts and fanouthosts fixtures in tests that don't need them

Signed-off-by: Saikrishna Arcot <sarcot@microsoft.com>

mssonicbld commented:
/azp run

saiarcot895 committed:
Merge remote-tracking branch 'origin/master' into reuse-existing-image-in-upgrade

saiarcot895 committed:
Reuse the target image as well

Signed-off-by: Saikrishna Arcot <sarcot@microsoft.com>

mssonicbld commented:
/azp run
